content/smyja/how-to-use-graphql-with-django.md

```diff
@@ -0,0 +1,436 @@
+---
+Title: "How To Use GraphQL With Django"
+Description: "A guide on how to use GraphQL with Django."
+DatePublished: "2022-06-05"
+Categories:
+  - "python"
+  - "web-development"
+  - "computer-science"
+Tags:
+  - "APIs"
+  - "Django"
+  - "Queries"
+CatalogContent:
+  - "python-for-programmers"
+  - "paths/full-stack-engineer-career-path"
+---
+
+_**Prerequisites:** Understanding of Python, the Command Line, and Django._  
+_**Versions:** Django 4.0.4, Python 3.8.10, virtualenv 20.15.1_
+
+### Introduction
+
+[GraphQL](https://www.codecademy.com/resources/docs/general/graphql) is a query language for [APIs](https://www.codecademy.com/resources/docs/general/api) and a runtime for fulfilling those queries with existing data. Unlike a REST API, GraphQL APIs do not require verbs (`PUT`, `POST`, `GET`, `PATCH`, and `DELETE`) for requests, nor do they need multiple endpoints. They have just one endpoint and making a query to that endpoint is all that's needed. 
+
+This tutorial will cover the creation of a CRUD (create, read, update, and delete) GraphQL API for a restaurant with Django.
+
+## Properties of GraphQL
+
+The following terms are often used when interacting with GraphQL. Knowing them can be helpful, though we won't be covering all of them in this tutorial.
+
+- **Schema**: Describes the functionality available to the client applications that connect to it.
+- **Query**: A schema type that represents the `GET` request and defines the operations that can be used for reading or fetching data.
+- **Nesting**: Queries can be nested inside of other queries.
+- **Mutation**: A schema type that defines the kind of operations that can be done to modify data.
+- **Subscription**: Notifies the client server in real time about updates to the data.
+- **Resolver**: Functions that return values for fields associated with existing schema types.
+
+## Step 1: setting up our virtual environment
+
+First, we are going to create and launch a virtual environment for our project with the `virtualenv` package (which can be [installed via `pip`](https://virtualenv.pypa.io/en/latest/installation.html#via-pip). While not necessary for starting a new Django project, working in separate environments is generally a best practice that mitigates conflicts between sites. Let's open a [terminal](https://www.codecademy.com/resources/docs/general/terminal) and create a new environment named `my_env` by running the following:
+
+```bash
+virtualenv my_env
+```
+
+Next, we will activate our new environment `my_env` with either of the following commands:
+
+```bash
+# Linux/macOS
+source my_env/bin/activate
+
+# Windows
+source/scripts/activate
+```
+
+Let's go to the next step.
+
+## Step 2: creating our Django project
+
+Next, if we haven't already, let's [install](https://docs.djangoproject.com/en/4.0/topics/install/#how-to-install-django) the `Django` package. Then, let's create a new project called `restaurant_graphql_api` and change into it:
```

Comment:
```suggestion
Next, if we haven't already, let's [install](https://docs.djangoproject.com/en/4.0/topics/install/#how-to-install-django) the `Django` package. 

Once we've done that, let's create a new project called `restaurant_graphql_api` and change into it:
```

Reply:
Just a clarification, as having these two distinct steps in the same paragraph might lead to some confusion.